src/core/utils/hooks/useOverflow.tsx

```diff
@@ -60,21 +70,26 @@ export const useOverflow = <T extends HTMLElement>(
     const availableWidth = containerRef.current.offsetWidth;
     const requiredWidth = containerRef.current.scrollWidth;
 
-    // hide items when there's no space available
-    if (availableWidth < requiredWidth && visibleCount > 1) {
-      setVisibleCount((count) => count - 1);
-      overflowBreakpoints.current.push(availableWidth);
-    }
-    // restore items when there's enough space again
-    else if (
-      overflowBreakpoints.current.length > 0 &&
-      availableWidth >
-        overflowBreakpoints.current[overflowBreakpoints.current.length - 1]
-    ) {
-      setVisibleCount((count) => count + 1);
-      overflowBreakpoints.current.pop();
+    if (availableWidth < requiredWidth) {
+      const avgItemWidth = requiredWidth / visibleCount;
+      const visibleItems = Math.floor(availableWidth / avgItemWidth);
+      setVisibleCount(visibleItems);
+    } else if (needsFullRerender.current) {
+      const childrenWidth = Array.from(containerRef.current.children).reduce(
+        (sum: number, child: HTMLElement) => sum + child.offsetWidth,
+        0,
+      );
+      const avgItemWidth = childrenWidth / visibleCount;
+      const visibleItems = Math.floor(availableWidth / avgItemWidth);
+      // Doubling the visible items to overflow the container. Just to be safe.
+      setVisibleCount(Math.min(items.length, visibleItems * 2));
     }
-  }, [containerWidth, visibleCount, disabled]);
+    needsFullRerender.current = false;
+  }, [containerWidth, visibleCount, disabled, items.length]);
```

Comment:
Just noticed `containerWidth` in deps array but it's not used

Reply:
Needed